libtbx/env_config.py

```diff
@@ -3182,9 +3182,75 @@ def get_boost_library_with_python_version(name, libpath):
         return name_version
   return name
 
+def check_libcpp():
+
+  #______________________________________________________________________________
+  # Skip if not Linux
+  # TODO: this might also work on macOS
+  #
+  if sys.platform.startswith("linux"):
+    return True
+  #
+  #------------------------------------------------------------------------------
+
+  import libtbx.env_config
+  from   libtbx             import easy_run
+  from   os                 import listdir
+
+  #______________________________________________________________________________
+  # get the libstd version that the python executable links against
+  #
+  python_libstdcxx_so = None
+  ldd_exec = easy_run.fully_buffered(command=f"ldd {sys.executable}")
+  for line in ldd_exec.stdout_lines:
+    if line.strip().startswith("libstdc++.so"):
+      python_libstdcxx_so = line.split()[0]
+      break
+
+  if python_libstdcxx_so is None:
+    return False
+  #
+  #------------------------------------------------------------------------------
+
+  #______________________________________________________________________________
+  # Go through all installed *.so's and ensure that they are linking to the same
+  # libstdc++ as python executable
+  #
+  lib_dir = join(
+      libtbx.env.lib_path._anchor._path, libtbx.env.lib_path.relocatable
+  )
+  libs = [join(lib_dir, name) for name in listdir(lib_dir) if name.endswith(".so")]
+  for lib in libs:
+    ldd_lib = easy_run.fully_buffered(command=f"ldd {lib}")
+    for line in ldd_lib.stdout_lines:
+      if line.strip().startswith("libstdc++.so"):
+        lib_libstdcxx_so = line.split()[0]
+        if lib_libstdcxx_so != python_libstdcxx_so:
+          raise SystemError(
+            f"FATAL: libstdc++.so mismatch\n"
+            f"sys.executable: {sys.executable}\n"
+            f"lib_file: {lib}"
+          )
+        break
+  #
+  #------------------------------------------------------------------------------
+
+  return True
+
+
 if (__name__ == "__main__"):
   if (len(sys.argv) == 2 and sys.argv[1] == "__libtbx_refresh__"):
     unpickle().refresh()
+
+    def is_true(x):
+      yes = {"1", "on", "true"}
+      return (x.strip().lower() in yes)
+
+    if is_true(os.environ.get("CCTBX_CHECK_LDD", "True")):
+      print("Checking that extension modules and python executabel link to same version of libstdc++")
+      print("Set CCTBX_CHECK_LDD=false to disable this step")
+      check_libcpp()
```

Comment:
We probably only need to check that `CCTBX_CHECK_LDD` is set, not the specific value. So maybe something like
```
if os.environ.get("CCTBX_CHECK_LDD", None) is not None:
```
instead of the `is_true` function.

Reply:
that will accept an empty string. Just use
```python
if os.getenv("CCTBX_CHECK_LDD"):
```

Also: *executab**le**

Reply:
and that (and your comment) inverses the logic, so make it `CCTBX_SKIP_LDD_CHECK`

Reply:
Personally I prefer to be more strict so that things like `CCTBX_CHECK_LDD=""` or `CCTBX_CHECK_LDD="off"`  won't accidentally trigger the check -- I know we can use `unset`, but it's easy to overlook a blank variable using simple `echo $CCTBX_CHECK_LDD` (i.e. there is no quick one-liner in bash to check for nullity).

Reply:
Wait, are we always running this check on linux or only when the environment variable is set? At the start of `def check_libcpp`, there is a check for `linux` that returns `True`.

Reply:
Oh, it is run regardless of the environment variable being set. But if it is set, it only runs if it's "1", "on", or "true" (case insensitive)?

Reply:
yes the earlier check was inverted -- thanks for catching that -- once that's fixed, the test is on by default on linux (unless it's turned off).